### PR TITLE
Sample shift

### PR DESCRIPTION
**Pull request by luojing1211**
Shift time samples at each channel. 
ps: just start this PR so that we have a place to discuss it.

### CODE REVIEW

**Comment by mhvk** [comment]
@luojing1211 - this looks like a good approach. Need to think a bit about how to keep the shifts, as some form of metadata.

Also, in reality we would of course want to shift 2x4-bit samples, but I think we can worry about that later!

File: baseband_tasks/tests/test_sampling.py [hunk 1]
```diff
@@ -611,3 +611,88 @@ def signal(self, ih):
...
+        shifter.seek(read_start)
+        shifted = shifter.read(read_num)
+        # Shifted should have 3 time samples of upstream data before the read
+        # points.
+        self.ih.seek(read_start - np.abs(np.min(shift)))
+        raw_data = self.ih.read(100)
+        for ii, sf in enumerate(shift - np.min(shift)): # normalize shift
+            assert np.all(shifted[:,ii,0] == raw_data[sf:sf + read_num,ii,0])
+            assert np.all(shifted[:,ii,1] == raw_data[sf:sf + read_num,ii,1])
+        # Test cross frame
```
Comment (luojing1211) [inline baseband_tasks/tests/test_sampling.py:684]: The cross-frame test fails. Since the result did not fetch the data from previous data.

**Comment by mhvk** [comment]
@luojing1211 - I think the problem was that my example and your code did not do the advanced index quite right. After a lot of reading and trial & error, I think I've solved it. I pushed what hopefully is a correct version...

**Comment by mhvk** [comment]
PINT seems to be failing...

**Comment by luojing1211** [comment]
> PINT seems to be failing...

Yes, it is the example notebook fail. I think David Kaplan is working on it.

**Comment by luojing1211** [comment]
What is the next step towards merging? Figure out the metadata?  Are there any more tests I should add to it?

**Comment by mhvk** [comment]
I think we don't need to wait for the meta-data: right now, this is similar to the other resampling shifts, in that it basically asks the user to keep track of what they have done.

I'll have another look at the code & tests...

**Comment by mhvk** [comment]
This looked good, but I realized the definition of the shift is in the opposite direction of `ShiftAndResample`. It took some time to figure out exactly what to do. Anyway, update now pushed. Will merge if tests pass.

**Comment by luojing1211** [comment]
Thank you. I guess the definition is now become: moving data back is a positive shift and moving data forward is negative. Is this right?

**Comment by mhvk** [comment]
Positive shift means data occurs at a later time, which means at a given time the data will come from earlier in the stream...

**Pull request merged**